xen/arch/arm/tee/ffa_partinfo.c

```diff
@@ -78,79 +81,95 @@ int32_t ffa_handle_partition_info_get(uint32_t w1, uint32_t w2, uint32_t w3,
     if ( w5 == FFA_PARTITION_INFO_GET_COUNT_FLAG &&
          ctx->guest_vers == FFA_VERSION_1_1 )
     {
+        ret = FFA_RET_OK;
+        *count = 0;
+
         if ( ffa_fw_support_fid(FFA_PARTITION_INFO_GET) )
-            return ffa_partition_info_get(w1, w2, w3, w4, w5, count, fpi_size);
-        else
-        {
-            *count = 0;
-            return FFA_RET_OK;
-        }
+            ret = ffa_partition_info_get(w1, w2, w3, w4, w5, count, size);
+
+        return ret;
     }
     if ( w5 )
         return FFA_RET_INVALID_PARAMETERS;
 
-    if ( !ffa_rx )
-        return FFA_RET_DENIED;
-
     if ( !spin_trylock(&ctx->rx_lock) )
         return FFA_RET_BUSY;
 
-    if ( !ffa_fw_support_fid(FFA_PARTITION_INFO_GET) )
-    {
-        if ( ctx->guest_vers == FFA_VERSION_1_0 )
-            *fpi_size = sizeof(struct ffa_partition_info_1_0);
-        else
-            *fpi_size = sizeof(struct ffa_partition_info_1_1);
-
-        *count = 0;
-        ret = FFA_RET_OK;
-        goto out;
-    }
-
-    if ( !ctx->page_count || !ctx->rx_is_free )
-        goto out;
-    spin_lock(&ffa_rx_buffer_lock);
-    ret = ffa_partition_info_get(w1, w2, w3, w4, w5, count, fpi_size);
-    if ( ret )
-        goto out_rx_buf_unlock;
     /*
-     * ffa_partition_info_get() succeeded so we now own the RX buffer we
-     * share with the SPMC. We must give it back using ffa_rx_release()
-     * once we've copied the content.
+     * If the guest is v1.0, he does not get back the entry size so we must
+     * use the v1.0 structure size in the destination buffer.
+     * Otherwise use the size of the highest version we support, here 1.1.
      */
-
     if ( ctx->guest_vers == FFA_VERSION_1_0 )
+        dst_size = sizeof(struct ffa_partition_info_1_0);
+    else
+        dst_size = sizeof(struct ffa_partition_info_1_1);
+
+    dst_buf = ctx->rx;
+
+    if ( ffa_fw_support_fid(FFA_PARTITION_INFO_GET) )
     {
-        size_t n;
-        struct ffa_partition_info_1_1 *src = ffa_rx;
-        struct ffa_partition_info_1_0 *dst = ctx->rx;
+        ret = FFA_RET_DENIED;
 
-        if ( ctx->page_count * FFA_PAGE_SIZE < *count * sizeof(*dst) )
-        {
-            ret = FFA_RET_NO_MEMORY;
-            goto out_rx_release;
-        }
+        if ( !ffa_rx )
+            goto out;
+
+        if ( !ctx->page_count || !ctx->rx_is_free )
+            goto out;
+
+        spin_lock(&ffa_rx_buffer_lock);
 
-        for ( n = 0; n < *count; n++ )
+        ret = ffa_partition_info_get(w1, w2, w3, w4, w5, &num_ffa_sp,
+                                     &src_size);
+
+        if ( ret )
+            goto out_rx_buf_unlock;
+
+        /*
+         * ffa_partition_info_get() succeeded so we now own the RX buffer we
+         * share with the SPMC. We must give it back using ffa_rx_release()
+         * once we've copied the content.
+         */
+
+        /* we cannot have a size lower than 1.0 structure */
+        if ( src_size < sizeof(struct ffa_partition_info_1_0) )
```

Comment:
Should we check and reject if `src_size` is larger than `dst_size`?

Reply:
no we must not and this is the point here. If the VM is 1.0 but SPMC is 1.1 src_size is larger than dst_size.

Reply:
I'm sorry, I meant the other way around. If `src_size` is smaller than `dst_size. I guess it can't happen today since SPMC must be at least at version 1.1, but that logic is disconnected from this function.

Reply:
I am not sure i get what you mean here: the guest is not giving us a dst_size we are choosing one.
Smallest destination size is 1.0 that i test here, if we get less we have an error tested here.
What case do you have in mind ?

Reply:
What if the guest is at version 1.1 and the SPMC is at version 1.0?
Then src_size would be larger than dst_size.

Reply:
We do not support SPMC being 1.0
And even in this case we are filling the rest with 0 with the memset which in 1.1 case will just set 0 UUIDs

Reply:
OK